components/script/dom/path2d.rs

```diff
@@ -48,14 +53,94 @@ impl Path2D {
+
+        // Step 3. If one or more of matrix's m11 element, m12 element, m21
+        // element, m22 element, m41 element, or m42 element are infinite or
+        // NaN, then return.
+        if !matrix.m11.is_finite() ||
+            !matrix.m12.is_finite() ||
+            !matrix.m21.is_finite() ||
+            !matrix.m22.is_finite() ||
+            !matrix.m31.is_finite() ||
+            !matrix.m32.is_finite()
+        {
+            return Ok(());
+        }
+
+        // Step 4. Create a copy of all the subpaths in path. Let c be this copy.
+        let mut c = other.segments();
+
+        // Step 5. Transform all the coordinates and lines in c by the transform matrix matrix.
+        let apply_matrix_transform =
+            |point: Point2D<f64>| -> Point2D<f64> { matrix.transform_point(point) };
+
+        for segment in c.0.elements_mut() {
+            match *segment {
+                PathEl::MoveTo(point) => {
+                    let transformed_point = apply_matrix_transform(Point2D::new(point.x, point.y));
+                    *segment = PathEl::MoveTo(Point::new(transformed_point.x, transformed_point.y));
+                },
+                PathEl::LineTo(point) => {
+                    let transformed_point = apply_matrix_transform(Point2D::new(point.x, point.y));
+                    *segment = PathEl::LineTo(Point::new(transformed_point.x, transformed_point.y));
+                },
+                PathEl::QuadTo(point, point_1) => {
+                    let transformed_point = apply_matrix_transform(Point2D::new(point.x, point.y));
+                    let transformed_point_1 =
+                        apply_matrix_transform(Point2D::new(point_1.x, point_1.y));
+                    *segment = PathEl::QuadTo(
+                        Point::new(transformed_point.x, transformed_point.y),
+                        Point::new(transformed_point_1.x, transformed_point_1.y),
+                    );
+                },
+                PathEl::CurveTo(point, point_1, point_2) => {
+                    let transformed_point = apply_matrix_transform(Point2D::new(point.x, point.y));
+                    let transformed_point_1 =
+                        apply_matrix_transform(Point2D::new(point_1.x, point_1.y));
+                    let transformed_point_2 =
+                        apply_matrix_transform(Point2D::new(point_2.x, point_2.y));
+                    *segment = PathEl::CurveTo(
+                        Point::new(transformed_point.x, transformed_point.y),
+                        Point::new(transformed_point_1.x, transformed_point_1.y),
+                        Point::new(transformed_point_2.x, transformed_point_2.y),
+                    );
+                },
+                PathEl::ClosePath => {},
+            }
+        }
+
+        // Step 6. Let (x, y) be the last point in the last subpath of c
+        let last_point = self.path.borrow_mut().last_point();
```

Comment:
Instead of doing borrow_mut multiple times, we should do it once and use this ref.